PROTOCOL.md

```diff
@@ -995,6 +1065,178 @@ Within the checkpoint, the `remove` struct does not contain the `stats` and `tag
+- Remaining rows in the V2 spec checkpoint refers the other actions mentioned [here](#checkpoints-1)
+- All the non-file actions i.e. all actions except [add and remove file](#Add-File-and-Remove-File)
+must be part of the v2 spec checkpoint itself.
+- A writer could choose to include the [add and remove file](#Add-File-and-Remove-File) action in the
+V2 spec Checkpoint or they could write the [add and remove file](#Add-File-and-Remove-File) actions in
+separate [sidecar files](#sidecar-files). These sidecar files will then be referenced in the V2 spec checkpoint.
+The sidecar files reside in the `_delta_log/_sidecars` directory.
+- A V2 spec Checkpoint could reference zero or more [sidecar file actions](#sidecar-file-information).
+
+Note: A V2 spec Checkpoint can either have all the [add and remove file](#Add-File-and-Remove-File) actions
+embedded inside itself or all of them should be in [sidecar files](#sidecar-files). Having partial
+add and remove file actions in V2 Checkpoint and partial entries in sidecar files is not allowed.
+
+After producing a V2 spec checkpoint, a writer can choose to embed some or all of the V2 spec checkpoint in
+the `_last_checkpoint` file, so that readers don't have to read the V2 Checkpoint.
+
+E.g. showing the content of V2 spec checkpoint:
+```
+{"checkpointMetadata":{"flavor":"flat","tags":{}}}
+{"metaData":{...}}
+{"protocol":{...}}
+{"txn":{"appId":"3ba13872-2d47-4e17-86a0-21afd2a22395","version":364475}}
+{"txn":{"appId":"3ae45b72-24e1-865a-a211-34987ae02f2a","version":4389}}
+{"sidecarFile":{"path":"3a0d65cd-4056-49b8-937b-95f9e3ee90e5.parquet","sizeInBytes":2341330,"modificationTime":1512909768000,"type":"fileaction","tags":{}}
+{"sidecarFile":{"path":"016ae953-37a9-438e-8683-9a9a4a79a395.parquet","sizeInBytes":8468120,"modificationTime":1512909848000,"type":"fileaction","tags":{}}
+```
+
+Another example of a v2 spec checkpoint without sidecars:
+```
+{"checkpointMetadata":{"flavor":"flat","tags":{}}}
+{"metaData":{...}}
+{"protocol":{...}}
+{"txn":{"appId":"3ba13872-2d47-4e17-86a0-21afd2a22395","version":364475}}
+{"add":{"path":"date=2017-12-10/part-000...c000.gz.parquet",...}
+{"add":{"path":"date=2017-12-09/part-000...c000.gz.parquet",...}
+{"remove":{"path":"date=2017-12-08/part-000...c000.gz.parquet",...}
+```
+
+#### V1 Spec
+
+V1 Spec do not support [sidecar files](#sidecar-files) and [checkpoint metadata](#checkpoint-metadata).
+These are flat checkpoints which contains all actions mentioned [here](#checkpoints-1).
+
+### Checkpoint Naming Scheme
+Delta supports following three checkpoint naming scheme:
+
+#### UUID-named checkpoint
+This naming scheme represents a [V2 spec checkpoint](#v2-spec) with following file name: `n.checkpoint.u.{json/parquet}`,
+where `u` is a UUID and `n` is the snapshot version that this checkpoint represents.
+The UUID-named checkpoints may be in JSON or parquet format. Since these are following [V2 spec](#v2-spec), so they must
```

Comment:
```suggestion
The UUID-named checkpoints may be in JSON or parquet format. Since these are following [V2 spec](#v2-spec), they must
```
or
```suggestion
The UUID-named checkpoints may be in JSON or parquet format. These are following [V2 spec](#v2-spec), so they must
```

Reply:
Adopting 1st suggestion.